### ISSUE TITLE
Ability to apply different indent styles to completions and snippets

### ISSUE THREAD

**Issue by jrappen** [sublimehq/Packages#131]
### Applying different [indent styles](https://en.wikipedia.org/wiki/Indent_style)

Should be possible by using a `*.tmPreferences` metadata file that holds whitespace settings that are for example set to either:

```xml
<key>shellVariables</key>
<array>
    <dict>
        <key>name</key><string>SPACE_OR_NEWLINE</string>
        <key>value</key><string><![CDATA[
]]></string>
    </dict>
</array>
```

or

```xml
<key>shellVariables</key>
<array>
    <dict>
        <key>name</key><string>SPACE_OR_NEWLINE</string>
        <key>value</key><string><![CDATA[ ]]></string>
    </dict>
</array>
```

and then used by snippets or completions like:

```json
    { "trigger": "my_class\tclass", "contents": "my_class()${SPACE_OR_NEWLINE}{\n\t${1:first line}\n\t${2:second line}\n}" }
```

---

I already merged (most) snippet files into completions file for easier handling in #128. I could add some commits or open a new PR if this is desired. As far as my understanding goes, this would need three settings for:

* in between `class` and `first line`
* in the indented block between lines like `first line` and `second line`
* in between `second line` and the end

in the example above.

---

I'd be willing to do this for all languages, however I'd like to have a basic structure set first as well as a list of styles that should be included.

**Comment by jrappen** [comment]
Any updates on this and/or #128?

**Comment by jrappen** [comment]
* https://github.com/jsiwek/WhitesmithsIndentStyle
* https://github.com/SublimeTextIssues/DefaultPackages/issues/61

**Comment by FichteFoll** [comment]
If there was better tmPreferences integration for variables such as these, things like https://github.com/SublimeText/AsciiDoc/pull/5/commits/c8ea5940987509c7e53d8e3f66f63c496c0083d8 could also become feasable. Currently those have a bit too high of barrier of entry and would basically require a plugin to walk users through the changes, imo.

**Comment by jrappen** [comment]
Still up for a challenge. Any updates?

**Comment by jrappen** [comment]
@wbond Maybe you could ask John to implement some improvements regarding the variables in those TextMate files (see FichteFoll's link) in the next build. I believe this would be a quick fix and require only minor changes? Especially the dropdown lists support would greatly enhance snippets and completions for all languages.

**Comment by wbond** [comment]
Jon and I have not yet discussed ideas surrounding the improvement of dealing with different indentation rules. I would not expect such enhancements to be part of the next dev build. In general, I think we may want a larger change to how indentation rules are implemented and how preferences may affect those.

@keith-hall [mentioned recently](https://github.com/SublimeTextIssues/Core/issues/1583#issuecomment-278970133) the idea of working on documenting the current behavior and features surrounding indentation. This would be helpful in us coming up with a more elegant solution for dealing with these sorts of issues moving forward.

**Comment by jrappen** [comment]
@wbond have you decided, yet?

---

linking @keith-hall's docs on indentation here, too: <https://forum.sublimetext.com/t/everything-you-n-ever-wanted-to-know-about-indentation-in-st3/26207>

**Comment by wbond** [comment]
No, my focus recently has been on performance issues related to default packages. Unfortunately indentation stuff is somewhat low priority, in the grand scheme of things.

**Comment by jrappen** [comment]
## Extended proposal

I'd like to extend:

* use five variables
  * WHITESPACE_BEFORE
  * WHITESPACE_INSIDE_FIRST
  * WHITESPACE_INSIDE
  * WHITESPACE_INSIDE_LAST
  * WHITESPACE_AFTER

One could argue that WHITESPACE_AFTER doesn't make sense cause it's the same for all styles...but just for clarity's sake to get my point across I included it here.

compare the usage:

```text
return_value func_name(params)${WHITESPACE_BEFORE}{${WHITESPACE_INSIDE_FIRST}func_a();${WHITESPACE_INSIDE}func_b();${WHITESPACE_INSIDE_LAST}}${WHITESPACE_AFTER}$0
```

for some fake code like (pipe indicates cursor pos for `$0`):

### output: Allman style

```text
return_value func_name(params)
{
    func_a();
    func_b();
}
|
```

### output: K-and-R style

```text
return_value func_name(params) {
    func_a();
    func_b();
}
|
```

### output: pico style

```text
return_value func_name(params)
{  func_a();
    func_b(); }
|
```

**Comment by rwols** [comment]
I like this proposal a lot; it mirrors the options for brace placement styles in clang-format.

**Comment by jrappen** [comment]
Implementation:

Using [`mdpopups`](https://github.com/facelessuser/sublime-markdown-popups) works great for feeding it a small code block to use [`sublime_plugin.ListInputHandler`](http://www.sublimetext.com/docs/3/api_reference.html#sublime_plugin.ListInputHandler) for displaying a preview for each indent style during the selection via the command palette.

**Comment by jeffreyscottgraham** [comment]
Is this plugin done and available somewhere?

**Comment by jrappen** [comment]
@jeffreyscottgraham mdpopups is available as a Sublime Text 3 dependency for your plug-in.

https://packagecontrol.io/docs/dependencies

https://github.com/facelessuser/sublime-markdown-popups

**Comment by jrappen** [comment]
@jeffreyscottgraham did you need help with the implementation on your side?

**Comment by jeffreyscottgraham** [comment]
Thanks no.

**Comment by FichteFoll** [comment]
Random thought I had the other day:

This would be made easier if snippets could also read variables from other locations, such as a view's settings, or if plugins could inject shellVariables (which are per selector), so they wouldn't have to generate tmPreferences files in some temporary location.
